### PR TITLE
IGNITE-14895 Reflection replaced with method handles in GridUnsafe.

### PR DESCRIPTION
**Pull request by ibessonov**
https://issues.apache.org/jira/browse/IGNITE-14895

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/util/GridUnsafe.java [hunk 1]
```diff
@@ -191,57 +197,59 @@ private GridUnsafe() {
...
-        catch (ReflectiveOperationException e) {
+        catch (Throwable e) {
             throw new RuntimeException("DirectByteBuffer#constructor is unavailable."
                 + FeatureChecker.JAVA_VER_SPECIFIC_WARN, e);
         }
     }
 
     /**
-     * Wraps a pointer to unmanaged memory with a direct byte buffer using a JavaNioAccess object.
+     * Wraps a pointer to unmanaged memory into direct byte buffer. Uses JavaNioAccess object.
```
Comment (sashapolo) [inline modules/core/src/main/java/org/apache/ignite/internal/util/GridUnsafe.java:229]:
```suggestion
     * Wraps a pointer to unmanaged memory into a direct byte buffer. Uses the JavaNioAccess object.
```

File: modules/core/src/main/java/org/apache/ignite/internal/util/GridUnsafe.java [hunk 2]
```diff
@@ -1587,17 +1595,31 @@ private static Object javaNioAccessObject() {
...
      */
-    private static Method newDirectBufferMethod(Object nioAccessObj) {
+    private static MethodHandle newDirectBufferMethodHandle(Object nioAccessObj) {
         try {
             Class<?> cls = nioAccessObj.getClass();
 
             Method mtd = cls.getMethod("newDirectByteBuffer", long.class, int.class, Object.class);
 
-            mtd.setAccessible(true);
+            AccessController.doPrivileged((PrivilegedExceptionAction<?>)() -> {
```
Comment (sashapolo) [inline modules/core/src/main/java/org/apache/ignite/internal/util/GridUnsafe.java:1604]: why is this needed now?

File: modules/core/src/main/java/org/apache/ignite/internal/util/GridUnsafe.java [hunk 3]
```diff
@@ -1587,17 +1595,31 @@ private static Object javaNioAccessObject() {
...
+                ByteBuffer.class,
+                Object.class,
+                long.class,
+                int.class,
+                Object.class
+            );
+
+            return MethodHandles.lookup()
+                .unreflect(mtd)
+                .asType(mtdType);
```
Comment (sashapolo) [inline modules/core/src/main/java/org/apache/ignite/internal/util/GridUnsafe.java:1620]: why is this line needed?

File: modules/core/src/main/java/org/apache/ignite/internal/util/GridUnsafe.java [hunk 4]
```diff
@@ -100,32 +103,35 @@
     /** {@link java.nio.Buffer#address} field offset. */
     private static final long DIRECT_BUF_ADDR_OFF = bufferAddressOffset();
 
+    /** Null object. */
+    private static final Object NULL_OBJ = null;
```
[the remark this answers is not in the document]
Comment (ibessonov) [reply modules/core/src/main/java/org/apache/ignite/internal/util/GridUnsafe.java:107]: I think **(Object)null** will confuse people more then this constant.

File: modules/core/src/main/java/org/apache/ignite/internal/util/GridUnsafe.java (hunk 2 above)
Comment (ibessonov) [reply modules/core/src/main/java/org/apache/ignite/internal/util/GridUnsafe.java:1604]: It's a good rule to execute all "security-sensitive" operations in privileged sections. Otherwise code may fail when security manager enabled.

File: modules/core/src/main/java/org/apache/ignite/internal/util/GridUnsafe.java (hunk 3 above)
Comment (ibessonov) [reply modules/core/src/main/java/org/apache/ignite/internal/util/GridUnsafe.java:1620]: You mean "asType"? For "invokeExact" to work properly

File: modules/core/src/main/java/org/apache/ignite/internal/util/GridUnsafe.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (sashapolo) [reply modules/core/src/main/java/org/apache/ignite/internal/util/GridUnsafe.java:107]: This class is called `GridUnsafe` and uses a ton of cryptic stuff, are you sure its the `(Object)null` that will confuse them?)))

File: modules/core/src/main/java/org/apache/ignite/internal/util/GridUnsafe.java (hunk 3 above)
Comment (ibessonov) [reply modules/core/src/main/java/org/apache/ignite/internal/util/GridUnsafe.java:1620]: Rechecked, code won't work without explicit type:
java.lang.invoke.WrongMethodTypeException: expected (,long,int,Object)ByteBuffer but found (Object,long,int,Object)ByteBuffer

File: modules/core/src/main/java/org/apache/ignite/internal/util/GridUnsafe.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (ibessonov) [reply modules/core/src/main/java/org/apache/ignite/internal/util/GridUnsafe.java:107]: I'd be confused by that))

**Review by sashapolo — APPROVED** [review]
